### PR TITLE
Kroppeb's refactor

### PR DESCRIPTION
**Pull request by Kroppeb**

### CODE REVIEW

File: processor/src/main/kotlin/com/thinkslynk/fabric/annotations/FabricProcessor.kt [hunk 1]
```diff
@@ -1,62 +1,121 @@
...
+            fun process(processor: Processor) {
+                if (processor in found)
+                    return
+                // if finder already removed, we got a cycle
+                if (!processing.add(processor)) {
+                    throw RuntimeException("Finders are cyclic. Cycle contains:$processor")
+                }
+
+                // true if a child errors
+                for (nextFinder in processor.requires) {
```
Comment (thinkslynk) [inline processor/src/main/kotlin/com/thinkslynk/fabric/annotations/FabricProcessor.kt:55]: A style comment, but I prefer the `.forEach` syntax; you yourself use it right after this.
```kotlin
processor.requires.forEach{process(it)}
```

File: processor/src/main/kotlin/com/thinkslynk/fabric/annotations/FabricProcessor.kt [hunk 2]
```diff
@@ -1,62 +1,121 @@
...
+
+        val processors = setupAllFinders()
+        private fun setupAllFinders(): List<Processor> {
+            val processing: MutableSet<Processor> = mutableSetOf()
+            val found: MutableSet<Processor> = LinkedHashSet()
+
+            // * marks a finder as processing
+            // * processes required processing
+            //      * if any are already marked they will error
+            fun process(processor: Processor) {
```
Comment (thinkslynk) [inline processor/src/main/kotlin/com/thinkslynk/fabric/annotations/FabricProcessor.kt:46]: I'd recommend moving this setup function else where and promoting the inner function:

```kotlin
class DependencyResolution{
    val processing: MutableSet<Processor> = mutableSetOf()
    val found: MutableSet<Processor> = LinkedHashSet()

    fun process(processor: Processor) { ... }

    fun retrieveOrderedProcessors(vararg generators: Generator): LinkedHashSet<Processor> {
        processing.clear()
        found.clear()

        generators.forEach { it.finders.forEach(::process) }
        return found
    }
}
```

File: processor/src/main/kotlin/com/thinkslynk/fabric/annotations/FabricProcessor.kt [hunk 3]
```diff
@@ -1,62 +1,121 @@
...
+            for (processor in processors) {
+                processingEnv.messager.info("running: $processor")
+                processor.run(roundEnv, processingEnv)
+            }
+
+            // process generators
+            for (generator in generators) {
+                processingEnv.messager.info("running: $generator")
+                generator.generate(path, processingEnv)
+            }
```
Comment (thinkslynk) [inline processor/src/main/kotlin/com/thinkslynk/fabric/annotations/FabricProcessor.kt:105]: I'm not going to point them all out, because again it's a style comment, but it's fewer characters to use forEach and is more in line with modern functional kotlin styles.

https://kotlinlang.org/docs/tutorials/kotlin-for-py/functional-programming.html#comprehensions

This style also allows for easy parallelization by simply adding `.parallelStream()` into the mix.

File: processor/src/main/kotlin/com/thinkslynk/fabric/annotations/builder/RegistryBuilder.kt [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    fun addDefinition(element: Definition) {
+        objectBuilder
+            .addProperty(
+                PropertySpec.builder(element.name, element.type)
+                    .initializer(element.initializer)
+                    .build()
+            )
+        registerBuilder
+            .addStatement(
+                """%T.register(%M, %T(%S, %P), %L)""",
```
Comment (thinkslynk) [inline processor/src/main/kotlin/com/thinkslynk/fabric/annotations/builder/RegistryBuilder.kt:39]: You can certainly do it this way, it is the way it's documented, but fStrings are not nearly as readable as a simple template string. E.g. a developer would have to find the documentation relevant to this library to know what `%T` even means. I view that as less maintainable than just:

```kotlin
"Registry.register(Registry.BLOCK, Identifier(\"${annotation.namespace}\", \"${annotation.name}\"),  $propName)"
```

(Which is what it was before)

File: processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/Iterable.kt [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package com.thinkslynk.fabric.annotations.extensions
+
+fun <T,S,R>Iterable<T>.crossFlatMap(extract:(T)->Iterable<S>, transform:(List<S>)->R):List<R>{
```
Comment (thinkslynk) [inline processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/Iterable.kt:3]: This could do with a refactor, or comments, or both. It's hard to follow.

File: processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/Iterable.kt [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    if(each.any{it.isEmpty()})
+        return emptyList()
+    val idx = IntArray(each.size)
+    val items = MutableList(each.size){each[it].first()}
+    var i = 0
+    while(true){
+        result.add(transform(items))
+
+        while(idx[i] + 1 == each[i].size){
+            if(++i == idx.size){
```
Comment (thinkslynk) [inline processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/Iterable.kt:15]: Prefix/postfix addition is confusing to a lot of developers. Let's just take one extra line to make it clear:

```kotlin
i +=1
if(i == idx.size){
    return result
}
```

File: processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/Iterable.kt [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+                return result
+            }
+        }
+        items[i] = each[i][++idx[i]]
```
Comment (thinkslynk) [inline processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/Iterable.kt:19]: Same here, move prefix addition to the line before.

File: processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/Iterable.kt [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+        while(i > 0){
+            idx[--i] = 0
```
Comment (thinkslynk) [inline processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/Iterable.kt:21]: Move prefix subtraction to line before

File: processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/Iterable.kt (hunk 5 above)
Comment (thinkslynk) [inline processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/Iterable.kt:3]: I took a stab at commenting it up and making it a little more understandable for other developers, feel free to iterate on this further:

```kotlin
fun <T,S,R>Iterable<T>.crossFlatMap(
    extract:(T)->Iterable<S>,
    transform:(List<S>)->R
):List<R>{
    val result = mutableListOf<R>()
    val each = map{extract(it).toList()}

    if(each.any{it.isEmpty()})
        return emptyList()

    // Create an array of counts, one for each element
    val idx = IntArray(each.size)

    // Create a list pointing to the first element in each sub list
    val items = MutableList(each.size){each[it].first()}
    while(true){
        var i = 0
        // Transform our cross section and store the result
        result.add(transform(items))

        // While our current counter is equal to its sublist size
        while(idx[i] + 1 == each[i].size){
            // Move on to the next list
            i += 1

            // If we're at the end of all our lists, return result
            if(i == idx.size){
                return result
            }
        }

        // Increase our count at the current index
        idx[i] += 1

        // Overwrite the item in our cross section
        items[i] = each[i][idx[i]]

        // Reverse iterate and clear
        (i-1 downTo 0).forEach {
            idx[it] = 0
            items[it] = each[it][0]
        }
    }
}
```

File: processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/Iterable.kt [hunk 9]
```diff
@@ -0,0 +1,25 @@
+package com.thinkslynk.fabric.annotations.extensions
+
+fun <T,S,R>Iterable<T>.crossFlatMap(extract:(T)->Iterable<S>, transform:(List<S>)->R):List<R>{
+    val result = mutableListOf<R>()
+    val each = map{extract(it).toList()}
+    if(each.any{it.isEmpty()})
+        return emptyList()
```
Comment (thinkslynk) [inline processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/Iterable.kt:7]: Why? (Not saying this isn't correct, just wondering why this is done. Would a filter be better? Or is the entire operation unsalvageable if even one sub list is empty?)

File: processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/Messager.kt [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+fun Messager.warn(s: String) = printMessage(Diagnostic.Kind.WARNING, s)
+fun Messager.warn(s: String, element: Element) = printMessage(Diagnostic.Kind.WARNING, s, element)
+
+fun Messager.error(s: String) = printMessage(Diagnostic.Kind.ERROR, s)
+fun Messager.error(s: String, element: Element) = printMessage(Diagnostic.Kind.ERROR, s, element)
+
+
+enum class LogLevel : Comparable<LogLevel>{
+    DEBUG, INFO, LOG, NORMAL
+}
```
Comment (thinkslynk) [inline processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/Messager.kt:34]: This file is fine for now; may want to switch to log4j at some point.

File: processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/TypeSpec.kt [hunk 11]
```diff
@@ -2,37 +2,33 @@ package com.thinkslynk.fabric.annotations.extensions
...
+        val targetClassName = element.simpleName.toString()
         val propName = formatName(targetClassName)
-        if (it.kind == ElementKind.METHOD) {
-            val methodElement = it as ExecutableElement
-            val returnType = methodElement.returnType
+        if (element.isMethod()) {
+            val returnType = element.returnType
 
-            b = b.addProperty(
+            this.addProperty(
```
Comment (thinkslynk) [inline processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/TypeSpec.kt:17]: I'm fine with / prefer this change, but double check that `this` is mutated by `addProperty`; sometimes builders return new copies with the modification, rather than mutating the current state. If that were the case, this would functionally do nothing because we're not capturing the mutated state. (That's why I did the ` b = b.bla()` stuff)

File: annotations/src/main/kotlin/com/thinkslynk/fabric/annotations/registry/RegisterBlockItem.kt [hunk 12]
```diff
@@ -3,7 +3,7 @@ package com.thinkslynk.fabric.annotations.registry
 @Retention(AnnotationRetention.SOURCE)
 @Target(AnnotationTarget.CLASS)
 annotation class RegisterBlockItem(
-        val namespace: String,
-        val name: String,
+        val namespace: String = "",
+        val name: String = "",
```
Comment (thinkslynk) [inline annotations/src/main/kotlin/com/thinkslynk/fabric/annotations/registry/RegisterBlockItem.kt:7]: Are these actually optional? And if so, how are you resolving them?

**Review by thinkslynk — CHANGES_REQUESTED** [review]

File: annotations/src/main/kotlin/com/thinkslynk/fabric/annotations/registry/RegisterBlockItem.kt (hunk 12 above)
Comment (Kroppeb) [reply annotations/src/main/kotlin/com/thinkslynk/fabric/annotations/registry/RegisterBlockItem.kt:7]: These are optional, yes, and if empty they will be set to the same value as the block annotation.

Comment (thinkslynk) [reply annotations/src/main/kotlin/com/thinkslynk/fabric/annotations/registry/RegisterBlockItem.kt:7]: Sounds good, just double checking.

File: processor/src/main/kotlin/com/thinkslynk/fabric/annotations/builder/RegistryBuilder.kt (hunk 4 above)
Comment (Kroppeb) [reply processor/src/main/kotlin/com/thinkslynk/fabric/annotations/builder/RegistryBuilder.kt:39]: `%T` (Type) also makes sure I'm actually passing a type there and will automatically add the import.
`%M` (Member) is the same but for Members
`%S` (String) adds `"` around (and escapes `$`)
`%P` (Parametrised String?) is like `%S` but without the escaping
`%N` (Name) Is for referencing a generated thing (by name) (And I should probably be using that instead of `%L` here)
`%L` (Literal) is for literally copying the string passed and is identical to `${stuff}`

I also prefer it as it makes code that can be easily split over multiple lines and doesn't require `\"` although I'm already using `"""` here anyway so I could just use `"`

File: processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/Iterable.kt (hunk 5 above)
Comment (Kroppeb) [reply processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/Iterable.kt:3]: Sure I'll add an explanation based on what I was thinking while I was creating this, and a docstring on what it does.

File: processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/Iterable.kt (hunk 9 above)
Comment (Kroppeb) [reply processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/Iterable.kt:7]: This code produces a cartesian product and applies a `map` operation on that. If one of the inputs of a cartesian product is the empty set than you'd expect the result to be empty too. 

Atm where this code is called, I'm sure that no input list is ever empty but this is what I'd expect if I have a case where it could be. I can accept throwing an exception in case, but I can't accept filtering out the empty lists first.
If one of the inputs would be dynamic I wouldn't expect my input list of 5 elements to suddenly drop to only 4 cause 1 of my inputs is empty, without me being able to infer which index I'm missing.

File: processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/TypeSpec.kt (hunk 11 above)
Comment (Kroppeb) [reply processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/TypeSpec.kt:17]: I'm sure that they are mutated as I've used them before and rely on it in other places of my code. Although I couldn't find a mention of them in the docs.
AFAIK almost all builders in java are mutable as that's their main point for existing. For building immutable objects.

Comment (thinkslynk) [reply processor/src/main/kotlin/com/thinkslynk/fabric/annotations/extensions/TypeSpec.kt:17]: Maybe it's common in java, not sure. In python I use `dataframe`s all the time, and those do not mutate state (I do a lot of pyspark.) I'd figure with kotlin's data classes, it would be common to mutate state with a `return this.copy(field=newVal)` so that you build through immutable objects. 

But point taken that you've tested this and it's fine. Just more wanted to explain why I did it the other way to begin with. If it functions properly this way, I'm all for it.
